synapse/handlers/space_summary.py

```diff
@@ -278,13 +286,19 @@ async def get_room_hierarchy(
         # world-readable)
         await self._auth.check_user_in_room_or_world_readable(room_id, requester)
 
-        # TODO Handle pulling previously persisted state by the pagination token.
+        # If this is continuing a previous session, pull the persisted data.
+        if from_token:
+            if from_token not in self._pagination_sessions:
+                raise SynapseError(400, "Unknown pagination token", Codes.INVALID_PARAM)
 
-        # the queue of rooms to process
-        room_queue = deque((_RoomQueueEntry(room_id, ()),))
+            # TODO Assert that other parameters have not changed.
+            (room_queue, processed_rooms) = self._pagination_sessions[from_token]
```

Comment:
I think this is going to go fairly badly wrong when we get two concurrent requests with the same `from_token`, and both end up popping entries from the same Deque. We might need a ResponseCache to provide locking? Or to copy the deque so that we can safely process both requests?

Reply:
A response cache sounds like it would be good regardless, since this could be a somewhat heavy calculation?